### DISCUSSION TITLE
Analyzing call graph for crazyflie-firmware

### DISCUSSION THREAD

**Discussion by leeminki02** [bitcraze/discussions#883] [Q&A] [answered]
Hello,
I want to ask how to analyze the call tree/call graph for the crazyflie firmware.

During research with crazyflie, I am working on some experimental crazyflie swarm implementations. I wanted to analyze if the modification I did affects and slows down the overall firmware execution (or any effects that cause the drone to calculate more, etc.). I have tried manually adding the DEBUG_PRINT function in some main functions in the firmware to get logs about function calls, but that was a bad idea since the print functions only returned one packet per millisecond.

So, I would like to know if I can analyze the crazyflie firmware to get call graphs or call trees during a simple flight.
If impossible, I am open to getting any other advice to find a way to gain information about how many times and in what order each function was called during the flight.

**Comment by krichardsson** [comment, marked as answer]
Hi!

Profiling/tracing code is generally a bit tricky in embedded systems. 

Often the debug port is used which requires a connected cable, this is not optimal if you actually want to fly ;-) On the other hand, real flying is maybe not that important from a profiling point of view? My guess is that more or less the same code will execute also if you fix the Crazyflie in a stand and "take off".

There are solutions where you connect external hardware (for instance https://www.lauterbach.com/) that can trace the execution without changing the code. There are other solutions that can trace in software only (for instance https://percepio.com/tracealyzer/), but then you have add stuff in the code (instrument the code) which will change the performance a bit, but this is usually not a problem. There might be some free tools, but these tools are often fairly expensive.

There might be some possibilities using GDB and the standard debugger, but I'm not sure what it supports in terms of counting function calls and similar.

A simple tool, that might be enough in this case, is to use the "Task dump" feature in the python client. You can find a button in the Console tab on the right side. When you click it (or set parameter system.taskDump to 1) the system will get a dump of the active tasks from FreeRTOS and dump it to the console log. Note that the stats are reset when you do a dump, and the numbers you get are accumulated since the last dump. This will enable you to see if some task is using more or less CPU time. Usually the "idle" task (that is all CPU time when no task is running) is a good number to check to see if your code is eating CPU or not.

Otherwise going down the route you started of adding counters or similar to your code is possible the simplest way forward. Just change your strategy to first write data to RAM (while you fly) and dump the result to the console later, perhaps triggered by a parameter value. 

There are also some counter/rate functionality in the Crazyflie that might be useful in logging rates in realtime, see the [statsCnt.h](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/utils/interface/statsCnt.h) file.

**Reply by leeminki02** [reply]
Well, Thanks a lot!
I've tried using call counter variables to store counter values in the firmware and printing as I click to print out the "Task Dump" button. It works really well, and I think it's just enough for my case. Of course, it would be better if I had any chance to trace code in real-time, but knowing the call counts would be just perfect for now! 
Thanks :)

**Reply by leeminki02** [reply]
@krichardsson By the way, to automate and control experiments more precisely, would there be any way to call the 'task dump' button function using cflib python code, instead of cfclient GUI?

**Reply by krichardsson** [reply]
You trigger the dump by setting the [parameter system.taskDump](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#systemtaskdump) to a non-zero value

**Comment by simondlevy** [comment]
This is a fascinating topic, and one that (along with neuromorphic sensing/control) I see Crazyflie as offering a unique opportunity for investigating.  Specifically, I am interested in applying concepts from "pure' programming languages like Haskell to the problem of sensing and flight-control.  More specifically, I am working on isolating the algorithmic parts of the Crazyflie firmware from the embbeded/RTOS code, by having the direction of call flow be solely from the latter to the former (as in Haskell's restriction that you cannot call monadic code from pure code).  In essence, the embedded parts can be written in C using FreeRTOS, and they call the algorithmic parts, which are implemented as header-only C++ classes.   

Although it probably wouldn't be practical to try to incorporate such modifications into the main firmware branch, I think that such an approach would help address some of the issues raised in this thread.
